src/common/serialization.hpp

```diff
@@ -17,59 +17,269 @@
-void serialize_desc(
-        serialization_stream_t &sstream, const reduction_desc_t &desc);
-void serialize_desc(
-        serialization_stream_t &sstream, const reorder_desc_t &desc);
-void serialize_desc(
-        serialization_stream_t &sstream, const resampling_desc_t &desc);
-void serialize_desc(serialization_stream_t &sstream, const rnn_desc_t &desc);
-void serialize_desc(serialization_stream_t &sstream, const sdpa_desc_t &desc);
-void serialize_desc(
-        serialization_stream_t &sstream, const shuffle_desc_t &desc);
-void serialize_desc(
-        serialization_stream_t &sstream, const softmax_desc_t &desc);
-void serialize_desc(serialization_stream_t &sstream, const sum_desc_t &desc);
-
-status_t serialize_desc(
-        serialization_stream_t &sstream, const op_desc_t *op_desc);
-
-} // namespace serialization
+
+#define DNNL_ASSERT_TRIVIALLY_SERIALIZABLE(cls) \
+    static_assert(serialization_stream_t::is_trivially_serialized<cls>::value, \
+            #cls " must be trivially serializable.")
+
+struct serialization_stream_t {
+    serialization_stream_t() = default;
+
+    template <typename Arg1, typename... Args>
+    serialization_stream_t(const Arg1 &a1, const Args &...args) {
+        append(a1, args...);
+    }
+
+    static serialization_stream_t from_data(std::vector<uint8_t> data) {
+        serialization_stream_t s;
+        s.data_ = std::move(data);
+        return s;
+    }
+
+    bool operator==(const serialization_stream_t &other) const {
+        return data_ == other.data_;
+    }
+
+#if defined(__cpp_lib_has_unique_object_representations) \
+        && __cpp_lib_has_unique_object_representations >= 201606L
+    template <typename T>
+    struct is_trivially_serialized {
+        static const bool value
+                = (std::has_unique_object_representations<T>::value
+                          || std::is_floating_point<T>::value)
+                && !(std::is_pointer<T>::value);
+    };
+
+#else
+    // Fallback for backward compatibility. As the structure layout should not
+    // change between c++ versions, compiling with c++17 will already verify the
+    // structures are valid for this use case.
+    template <typename T>
+    struct is_trivially_serialized {
+        static const bool value = std::is_trivially_copyable<T>::value
+                && !(std::is_pointer<T>::value);
+    };
+#endif
+
+    template <typename T>
+    struct has_serialize_t {
+        using yes_t = uint8_t;
+        using no_t = uint16_t;
+
+        template <typename U>
+        static yes_t test(utils::enable_if_t<
+                std::is_same<decltype(&U::serialize),
+                        void (U::*)(serialization_stream_t &) const>::value,
+                bool>);
+        template <typename U>
+        static no_t test(...);
+
+        static const bool value = (sizeof(test<T>(0)) == sizeof(yes_t));
+    };
+
+    // Append helper function for structures with the member function
```

Comment:
QQ: is `append` supposed to be public? Or should a ctor be the only user-facing?

Reply:
Yes, it's supposed to be public as it's used as `sstream.append(...)` in many places.
I put some suggestions to unify API in the description (e.g. switch to uniform `serialize(sstream, obj)` usage) but I'd left it for the future.